---
title: |
  | _Ambrosia_ beetle RNAseq --
  | Overview of results so far
pagetitle: "Ambrosia RNAseq results"
author: "Jelmer Poelstra (poelstra.1@osu.edu), MCIC Wooster, OSU"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cerulean
    highlight: kate
    toc: true
    toc_float: true
    fig_caption: true
    anchor_sections: true
    df_print: kable
    css: html_page.css
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  eval = TRUE,
  cache = FALSE,
  warning = FALSE,
  message = FALSE,
  out.width = "80%"
)
knitr::opts_knit$set(root.dir = here::here())

#bibliography: compare.bib
```

```{r}
## SETTINGS

## Gene selection
gene_selection <- "nonfungal"
#gene_selection <- "fungal"
#gene_selection <- "insect"

## Thresholds
p_tres <- 0.05
mean_tres <- 1
lfc_tres <- 0.5
```

```{r}
#TODO - USE LFC-THRESHOLD WITH S-VALUE? (USED S-VAL THRESHOLD OF 0.005 W/ RANJANA)
#TODO - GO - VERY MANY GO CATS PER GENE...
#TODO - GO - MISSING GENE LENGTHS!
```

```{r}
## Load packages
if (!"pacman" %in% installed.packages()) install.packages("pacman")
if (!"BiocManager" %in% installed.packages()) install.packages("BiocManager")
if (!"DESeq2" %in% installed.packages()) BiocManager::install("DESeq2")
packages <- c("tidyverse",
              "DESeq2", "goseq",
              "glue", "here", 
              "pheatmap", "ggrepel", "plotly", "patchwork", 
              "VennDiagram", "RColorBrewer", "colorspace",
              "DT", "kableExtra")
pacman::p_load(char = packages, install = TRUE)
```

```{r}
## Source script with functions
source(here("scripts/DE_funs.R"))
source(here("scripts/GO_funs.R"))
source(here("scripts/report_funs.R"))
```

```{r}
## Define input files
kallisto_dir <- here("results/kallisto/beetle")

dds_file <- here(kallisto_dir, "deseq_object.rds")

tax_file <- here("results/trinotate/blast_tax.txt")
annot_file <- here("results/trinotate/blast_res.txt")

GO_map_file <- here("results/GO/GO_map.txt")
gene_len_file <- here("results/GO/gene_lengths.txt")
```

```{r}
## Settings
## Contrast (comp) setting
complist <- list(c("FMy", "EMySc"), c("FMy", "FMySc"), c("EMySc", "FMySc"))
compvec <- map_chr(complist, function(x) paste0(x, collapse = " v. "))

## Plot settings
col_fac <- "tissue"
theme_set(theme_bw())
```

```{r}
## Load and prepare input data

## Load taxonomic and gene info
tax <- read_tsv(tax_file, show_col_types = FALSE)
annot <- read_tsv(annot_file, show_col_types = FALSE)

## Import DESeq2 object
dds_raw <- readRDS(dds_file)                                  # 60,751 genes
colData(dds_raw)$tissue <- factor(colData(dds_raw)$tissue)

## SUBSET DESEQ OBJECT ACCORDING TO TAXONOMY
fungus_genes <- tax %>% filter(taxlev2 == "Fungi") %>% pull(gene_id) %>% unique()
insect_genes <- tax %>% filter(taxlev6 == "Insecta") %>% pull(gene_id) %>% unique()
nonfungus_genes <- tax %>% filter(taxlev2 != "Fungi") %>% pull(gene_id) %>% unique()

if (gene_selection == "fungal") dds <- dds_raw[fungus_genes, ] # 6,894 genes
if (gene_selection == "insect") dds <- dds_raw[insect_genes, ] # 4,441 genes
if (gene_selection == "nonfungal") dds <- dds_raw[!(row.names(dds_raw) %in% fungus_genes), ] # 53,857 genes 

## Extract metadata
meta <- colData(dds) %>% as.data.frame()

## Get normalized count matrix
dds <- estimateSizeFactors(dds)
counts_raw <- dds@assays@data$counts
count_mat <- sweep(counts_raw, 2, sizeFactors(dds), "/")

## Get a long-format count df with metadata
count_df <- as.data.frame(count_mat) %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(-gene_id, names_to = "id_short", values_to = "count") %>%
  left_join(meta, by = "id_short")

## Gene lengths and GO map for GO analysis
GO_map <- read_tsv(GO_map_file) %>% as.data.frame()     # goseq fails with a tibble
gene_lens <- read_tsv(gene_len_file)
```

```{r}
## Tissue colors
fac_levs <- unique(meta[[col_fac]])
fac_cols <- qualitative_hcl(length(fac_levs), palette = "Dark 3")
names(fac_cols) <- fac_levs

## Contrast colors
compcols <- brewer.pal(length(compvec), name = "Dark2")
```

```{r}
## DE analysis
design(dds) <- formula(~ tissue)
dds <- DESeq(dds)

## Get DE results for each pairwise comparison
DE <- map_dfr(.x = complist, .f = get_DE,
              dds = dds, count_df = count_df, annot = annot,
              p_tres = 0.05, mean_tres = mean_tres, lfc_tres = lfc_tres) %>%
  mutate(contrast = sub("_", " v. ", contrast))

## Shrunken LFC values + LFC test
# DE_shrink <- map_dfr(complist, shrink_lfc, dds = dds, fac = "tissue")
```

<br>

----

## Introductory notes

### Overview of the methods

- FASTQ QC with `FastQC` and `MultiQC`
- FASTQ read correction with `rcorrector`
- FASTQ read trimming and filtering with `Trimmomatic`
- _As a first step in excluding reads of fungal origin_,
  reads were mapped to 25 fungal genomes with `STAR` and reads that successfully
  mapped to one or more of these genomes were excluded.
  The fungal reference genomes were the same as used in
  [Blaz et al. 2019](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6463014/),
  see their Table S2.
- Transcriptome assembly with `Trinity`
- Transcriptome assembly QC with `TrinityStats`, `BUSCO`, `rnaQUAST`,
  and `Detonate`
- Transcriptome annotation with `Trinotate`,
  which in turn incorporates `TransDecoder`, `BLASTX`, `BLASTP`, `SignalP`, `TMHMM`,
  `RNAMMER`, `HMMSCAN`, and `SwissProt` and `Pfam` databases.
- Read quantification with the pseudo-alignment approach `Kallisto`
- _As a second step in excluding reads of fungal origin_,
  `BLASTP` and `BLASTX` results as part of the `Trinotate` pipeline were used --
  see below for more details.
- Differential expression analysis with the R/Bioconductor package `DESeq2`.
  P-values were adjusted for multiple testing with the Benjamini-Hochberg correction,
  and genes were only considered significant if the adjusted p-value was
  below `r p_tres`, the mean expresssion level for one of the groups was at least
  `r mean_tres`, and the log2-fold change in expression level was at least
  `r lfc_tres`.
- Gene Ontology analysis with the R/Bioconductor package `goseq`
  (Using GO annotations from the `Trinotate` results).
  P-values were adjusted for multiple testing with the Benjamini-Hochberg correction,
  and categories were only considered significant if the adjusted p-value was
  below 0.05.

### Genes included in this report

```{r, results='asis'}
if (gene_selection == "insect") {
  sel_descript <- "only genes with a closest `BLAST` match of insect origin (class _Insecta_) were included"
} else if (gene_selection == "fungal") {
  sel_descript <- "only genes with a closest `BLAST` match of fungal origin were included"
} else if (gene_selection == "nonfungal") {
  sel_descript <- "genes with a closest `BLAST` match of fungal origin were _excluded_"
} else {
  stop("INVALID GENE SELECTION")
}
```

**For this report, `r sel_descript`.**

**Total number of genes included in this report: `r nrow(dds)`**
**(out of a total of `r nrow(dds_raw)` in the entire transcriptome).**

### Tissues

- Only mycangia -- `FMy`
- Mycangia + scutellum ("flap") -- `FMySC`
- Early development mycangia + scutellum ("flap") -- `EMySc`
  
### Abbreviations

- _DEG_ -- Differentially Expressed Gene
- _padj_ -- Adjusted p-value (Adjusted for multiple testing using the Benjamini-Hochberg correction)
- _LFC_ -- Log2-fold change in mean expression level between two tissues

<br>

----

## PCA {.tabset .tabset-fade .tabset-pills}

<br>

```{r}
pca_res <- plotPCA(varianceStabilizingTransformation(dds, blind = TRUE),
                   intgroup = "tissue", returnData = TRUE)
pct_var <- round(100 * attr(pca_res, "percentVar"), 2)
```

### Without sample labels

```{r}
## Plot w/o sample labels
ggplot(pca_res) +
  aes(x = PC1, y = PC2, fill = tissue) +
  geom_point(size = 5, shape = 21, color = "grey20") +
  scale_fill_manual(values = fac_cols) +
  xlab(paste0("PC1: ", pct_var[1], "% of variance")) +
  ylab(paste0("PC2: ", pct_var[2], "% of variance"))
```

<br>

----

### With sample labels

```{r}
## Plot w/ sample labels
ggplot(pca_res) +
  aes(x = PC1, y = PC2, fill = tissue) +
  geom_point(size = 4, shape = 21, color = "grey20") +
  scale_fill_manual(values = fac_cols) +
  geom_text_repel(aes(label = name)) +
  xlab(paste0("PC1: ", pct_var[1], "% of variance")) +
  ylab(paste0("PC2: ", pct_var[2], "% of variance"))
```

<br>

----

## DEG tables {.tabset .tabset-fade .tabset-pills}

```{r}
degtab_cap <- glue("mean_A / mean_B: mean count in group A / group B")
```

Out of `r nrow(count_mat)` genes, the following numbers are significantly
differentially expressed (adjusted p-value < 0.05) for each of the pairwise
comparisons:

```{r}
DE %>%
  group_by(contrast) %>%
  summarize(`nr of DEGs` = sum(isDE, na.rm = TRUE)) %>%
  make_kable()
```

<br>

### `r compvec[1]`

<br>

```{r}
DE %>%
  filter(contrast == compvec[1], isDE == TRUE) %>%
  select(mean_A, mean_B, lfc, padj, gene_name, description, gene_id) %>%
  mutate(description = str_trunc(description, width = 60)) %>% 
  make_dt(caption = degtab_cap)
```

<br>

----

### `r compvec[2]`

<br>

```{r}
DE %>%
  filter(contrast == compvec[2], isDE == TRUE) %>%
  select(mean_A, mean_B, lfc, padj, gene_name, description, gene_id) %>%
  mutate(description = str_trunc(description, width = 60)) %>% 
  make_dt(caption = degtab_cap)
```

<br>

----

### `r compvec[3]`

<br>

```{r}
DE %>%
  filter(contrast == compvec[3], isDE == TRUE) %>%
  select(mean_A, mean_B, lfc, padj, gene_name, description, gene_id) %>%
  mutate(description = str_trunc(description, width = 60)) %>% 
  make_dt(caption = degtab_cap)
```

<br>

----

### DEGs in all comparisons

The following genes were differentially expressed in all 3 pairwise comparisons
(the values in the contrast columns are adjusted p-values):

```{r}
DE_in_all <- DE %>%
  select(isDE, contrast, padj, gene_id, gene_name, description) %>%
  group_by(gene_id) %>%
  filter(sum(isDE) == length(compvec)) %>% 
  pivot_wider(values_from = padj, names_from = contrast) %>%
  select(-isDE)
```

```{r}
DE_in_all %>% 
  mutate(description = str_trunc(description, width = 60)) %>% 
  make_dt()
```

<br>

----

## Volcano plots {.tabset .tabset-fade .tabset-pills}

### Static plot

```{r}
pvolc(DE, cols = compcols)
```

### Interactive plot

```{r}
pvolc(DE, cols = compcols, interactive = TRUE)
```

<br>

----

## MA plots {.tabset .tabset-fade .tabset-pills}

DEGs are in blue.

### Static plot

```{r}
pMA(DE %>% filter(mean >= 1), compvec)

## With shrunken LFCs
# pMA(DE_shrink, compvec, rm_padj_na = FALSE)
# I don't think the shrunken LFC works so well here as there are so many DEGs!
```

### Interactive plot

```{r}
if (nrow(DE) < 100000) pMA(DE %>% filter(mean >= 1), compvec, interactive = TRUE)
```

```{r, results='asis'}
if (nrow(DE) >= 100000) cat("Too many genes for an interactive plot.\n")
```

<br>

----

## Heatmaps

```{r}
## Color list for heatmaps
fac_col_list <- list(fac_cols)
names(fac_col_list) <- "tissue"
```

Values displayed by colors are log10-transformed normalized counts.

### Top-100 most highly expressed genes

Here, genes are not clustered by overall expression pattern similarity
like in the other heatmaps, but just sorted by overall abundance.

```{r}
hi_expr <- names(sort(rowSums(counts_raw), decreasing = TRUE)[1:100])
```

```{r}
pheat(hi_expr, count_mat, meta,
      groups = "tissue",
      show_rownames = FALSE, cluster_rows = FALSE,
      annotation_colors = fac_col_list)
```

### Top-100 DEGs {.tabset .tabset-fade .tabset-pills}

#### `r compvec[1]`

```{r}
fDE <- DE %>% filter(contrast == compvec[1], isDE == TRUE)
fgenes <- fDE$gene_id[1:min(100, nrow(fDE))]
```

```{r}
pheat(fgenes, count_mat, meta,
      show_rownames = FALSE, annotation_colors = fac_col_list)
```

<br>

----

#### `r compvec[2]`

```{r}
fDE <- DE %>% filter(contrast == compvec[2], isDE == TRUE)
fgenes <- fDE$gene_id[1:min(100, nrow(fDE))]
```

```{r}
pheat(fgenes, count_mat, meta,
      show_rownames = FALSE, annotation_colors = fac_col_list)
```

<br>

----

#### `r compvec[3]`

```{r}
fDE <- DE %>% filter(contrast == compvec[3], isDE == TRUE)
fgenes <- fDE$gene_id[1:min(100, nrow(fDE))]
```

```{r}
pheat(fgenes, count_mat, meta,
      show_rownames = FALSE, annotation_colors = fac_col_list)
```

<br>

----

## Top-12 DE boxplots {.tabset .tabset-fade .tabset-pills}

<br>

### `r compvec[1]`

```{r}
fDE <- DE %>% filter(contrast == compvec[1], isDE == TRUE)
```

```{r}
## One plot per gene
#walk(DE_fE$gene_id[1:10], pbox, count_df, annot, fac_cols)
```

```{r}
p4box(fDE$gene_id[1:4], count_df, annot, fac_cols)
p4box(fDE$gene_id[5:8], count_df, annot, fac_cols)
p4box(fDE$gene_id[9:12], count_df, annot, fac_cols)
```

<br>

----

### `r compvec[2]`

```{r}
fDE <- DE %>% filter(contrast == compvec[2], isDE == TRUE)
```

```{r}
p4box(fDE$gene_id[1:4], count_df, annot, fac_cols)
p4box(fDE$gene_id[5:8], count_df, annot, fac_cols)
p4box(fDE$gene_id[9:12], count_df, annot, fac_cols)
```

<br>

----

### `r compvec[3]`

```{r}
# **Note that for all of these genes, the differences between FMy and the other**
# **two are much larger than for EMySC v. FMySc., even though these are the most**
# **significant genes for the latter pairwise comparison (!)**.
```

```{r}
fDE <- DE %>% filter(contrast == compvec[3], isDE == TRUE)
```

```{r}
p4box(fDE$gene_id[1:4], count_df, annot, fac_cols)
p4box(fDE$gene_id[5:8], count_df, annot, fac_cols)
p4box(fDE$gene_id[9:12], count_df, annot, fac_cols)
```

<br>

----

## Venn diagrams of DEG overlap {.tabset .tabset-fade .tabset-pills}

<br>

### All DEGs

Among all significant DEGs for each comparison,
how many genes overlap?

```{r, results='hide'}
flist <- list(filter(DE, contrast == compvec[1], isDE == TRUE) %>% pull(gene_id),
              filter(DE, contrast == compvec[2], isDE == TRUE) %>% pull(gene_id),
              filter(DE, contrast == compvec[3], isDE == TRUE) %>% pull(gene_id))
names(flist) <- compvec

p <- venn.diagram(x = flist, fill = compcols,
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  margin = 0.1, cat.cex = 0.9, cat.fontface = "bold")
grid::grid.draw(p)
```

### Top-100 DEGs

Among the top-100 most significant DEGs for each comparison,
how many genes overlap?

```{r, results='hide'}
flist <- list(filter(DE, contrast == compvec[1], isDE == TRUE) %>% pull(gene_id) %>% head(100),
              filter(DE, contrast == compvec[2], isDE == TRUE) %>% pull(gene_id) %>% head(100),
              filter(DE, contrast == compvec[3], isDE == TRUE) %>% pull(gene_id) %>% head(100))
names(flist) <- compvec

p <- venn.diagram(x = flist, fill = compcols,
                  filename = NULL, output = TRUE, disable.logging = TRUE,
                  margin = 0.1, cat.cex = 0.9, cat.fontface = "bold")
grid::grid.draw(p)
```

<br>

----

## Gene Ontology

<br>

```{r}
GO_res <- map_dfr(.x = unique(DE$contrast),
                  .f = GO_wrap,
                  DE, GO_map, gene_lens) %>%
  filter(padj < 0.05)
```

### Plot

- Only GO categories that are significantly enriched in at least one comparison
  are shown.
  
- Grey (blank) tiles indicate that the focal GO category was not significantly
 enriched for the focal pairwise comparison.

- Numbers in the tiles indicate the number of DEGs in that category.

- The plot is divided into the three sub-ontologies:
  - `BP` = Biological Process
  - `CC` = Cellular Component
  - `MF` = Molecular Function

```{r, fig.height = 10, out.width="100%"}
if (gene_selection == "nonfungal") {
  ylabsize <- 5
  cntlabsize <- 1.4
} else {
  ylabsize <- 7
  cntlabsize <- 2
}

prep_goplot(GO_res, contrasts = compvec) %>%
  goplot(ylabsize = ylabsize, cntlabsize = cntlabsize, repeat_labs = TRUE)
```

### Tables {.tabset .tabset-fade .tabset-pills}

When a table below is empty, there were no significantly enriched GO categories
among the DEGs for the focal pairwise comparison.

```{r}
gotab_cap <- glue("p: raw p-value; padj: adjusted p-value;
                  numDEInCat: number of DEGs in this GO category;
                  numInCat: total number of genes in this GO category")
```

#### `r compvec[1]` 

<br>

```{r}
GO_res %>% filter(contrast == compvec[1]) %>% make_dt(caption = gotab_cap)
```

#### `r compvec[2]` {.tabset .tabset-fade .tabset-pills}

<br>

```{r}
GO_res %>% filter(contrast == compvec[2]) %>% make_dt(caption = gotab_cap)
```

#### `r compvec[3]` {.tabset .tabset-fade .tabset-pills}

<br>

```{r}
GO_res %>% filter(contrast == compvec[3]) %>% make_dt(caption = gotab_cap)
```

<br>
